Pull request2

DanielDuchnicki opened the pull request:
- Wydziel wszystkie klasy odpowiadające za obsługę konsoli (z wyjątkiem stepów - to tylko naiwna implementacja) do osobnego miejsca. Najlepiej do nowego projektu - **FIXED**

- Napisałeś, że dodałeś testy tam gdzie mogłeś. Napisz proszę, dlaczego nie mogłeś dodać testów w innych przypadkach i będziemy to naprawiać, jeśli będzie warto ;) - **Chodziło raczej o metody, które są tymczasową implementacją stepów**

Hotel system

- Init nie powinien być częścią tej klasy, zrób Init używając AddHotel z zewnątrz. - **FIXED**

- Lista Hoteli nie powinna być dostępna z zewnątrz skoro chcesz mieć kontrolę nad dodawaniem hoteli - **FIXED**

- W testach możesz użyć asserta na całą kolekcję:  hotel.ReservationSteps.Should().BeEquivalentTo(reservationSteps) - **FIXED**

- Masz nazwy klas, które nazywają się tak jak nazwy folderów. Przez to musisz podawać typy klas zawierając całą ścieżkę np. List<ReservationSteps.ReservationSteps. Enum mógłby się nazwyać ReservationStepType i nie byłby tego problemu. Namespace Hotels można zamienić na Hotels i też znikinie problem. Pokombinuj sobie - **FIXED**

DanielDuchnicki committed:
Moving console user interface methods to the new project and some other changes.

michalborowski reviewed (commented):
Mógłbyś w projekcie testowym zmienić wersje paczek na:

"FluentAssertions" version="5.3.0" targetFramework="net452" />
"NUnit" version="3.11.0" targetFramework="net452" />

Pozwoli mi to łatwiej przeglądać w pracy ;D akturat nie mamy wersji, których użyłeś w internal repo

michalborowski commented on HotelReservation/ConsoleUserInterface/SystemMain.cs:
```diff
@@ -28,10 +53,11 @@ public static int Menu(HotelSystem hotelSystem, StepsExecutor stepsExecutor)
+                    var hotelId = SelectHotel();
                     Console.Clear();
-                    if (hotelId >= 1000)
+                    if (hotelId >= 1000 & hotelId < hotelSystem.LastHotelId)
```
Wiedza, że ID zaczyna się od 1000 to nie jest do końca coś, czego powinna być świadoma warstwa interfejsu. GetHotelReservationSteps(hotelId) może zwrócić błąd lub wyjątek jeśli nie ma hotelu z takim ID, jakiego szukasz

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/Hotels/HotelSystem.cs:
```diff
@@ -0,0 +1,29 @@
+    public class HotelSystem
+    {
+        private readonly List<Hotel> _hotels = new List<Hotel>();
+        public int LastHotelId { get; private set; } = 1000;
```
Wygląda na to, że nie posiadając żadnego hotelu dalej dostaje jakieś LastHotelId, które niczemu nie odpowiada

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/Hotels/HotelSystem.cs:
```diff
@@ -0,0 +1,29 @@
+            LastHotelId++;
+        }
+
+        public List<Hotel> GetHotels()
```
Równie dobrze _hotels może być public ;) Mogę zrobić GetHotels().Add(...) i to zostanie w systemie. Jeśli nie chcesz, żeby ktoś edytował listę (a przynajmniej nie tak łatwo), to zwróć np. ReadOnlyCollection.

DanielDuchnicki replied:
Fixed - zwraca ReadOnlyCollection

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepsExecutor.cs:
```diff
@@ -7,19 +7,19 @@ namespace HotelReservation.ReservationSteps
-                    {ReservationSteps.ReservationProcess, new ReservationStartProcess() },
-                    {ReservationSteps.SendingMailProcess, new SendingMailProcess() },
-                    {ReservationSteps.PaymentProcess, new PaymentProcess() }
+                    {ReservationStepType.ReservationProcess, new ReservationStartProcess() },
```
Kod się już trochę uprościł, więc wróćmy do tematu factory, którego kiedyś było dużo w projekcie

Executor wykonuje kroki zawsze tak samo (i dobrze). Nie powinien się on więc zmieniać w sytuacji, gdy dodajemy nowy krok. Warto by było wydelegować tworzenie intancji dla danego typu do osobnej klasy.
Drugą sprawą, która póki co ma małe znaczenie, jest tworzenie wszystkich instancji zanim ich potrzebujesz. Warto tworzyć takie obiekty tuż przed Execute() zmiast trzymać na zawsze w pamięci.

DanielDuchnicki replied:
Fixed - instancje tworzone w klasie StepFactory i zwracane przez metodę CreateInstance()

michalborowski replied:
> Drugą sprawą, która póki co ma małe znaczenie, jest tworzenie wszystkich instancji zanim ich potrzebujesz. Warto tworzyć takie obiekty tuż przed Execute() zmiast trzymać na zawsze w pamięci.

Zrobiłeś właśnie tak, że inicjalizujesz wszystkie stepy. Daj znać, czy to, co napisałem jest niezrozumiałe, czy celowo ominąłeś sugestię

DanielDuchnicki replied:
> > Drugą sprawą, która póki co ma małe znaczenie, jest tworzenie wszystkich instancji zanim ich potrzebujesz. Warto tworzyć takie obiekty tuż przed Execute() zmiast trzymać na zawsze w pamięci.
> 
> Zrobiłeś właśnie tak, że inicjalizujesz wszystkie stepy. Daj znać, czy to, co napisałem jest niezrozumiałe, czy celowo ominąłeś sugestię

Fixed - pominąłem, bo zrozumiałem, że skoro nie jest na tę chwilę ważne, to nie powinienem się tym zajmować :) przy okazji pytanie - co myślisz o używaniu "Lazy"?

michalborowski commented on HotelReservation/HotelReservationTests/Hotels/HotelSystemTests.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace HotelReservationTests.Hotels
+{
+    [TestFixture]
```
Możesz wydzielić wspólny setup do metody z atrybutem [SetUp] i tam stworzyć _subject = new HotelSystem()

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservationTests/Hotels/HotelSystemTests.cs:
```diff
@@ -0,0 +1,53 @@
+namespace HotelReservationTests.Hotels
+{
+    [TestFixture]
+    public class HotelSystemTests
```
Brakuje testów na:
* LastHotelId - a to publiczny interfejs. Ta propercja wg mnie jest niepotrzebna ale w momencie jej dodawania powinien powstać test
* GetHotelReservationSteps kiedy podany Id nie istnieje

DanielDuchnicki replied:
Fixed

michalborowski replied:
ShouldLastHotelIdBe0WhenNoHotelsAdded
ID jest numerem - 0 to też numer. Taka knonwencja jest bardzo niebezpieczna i prędzej czy później w projekcie gdzie jest kilka osób ktoś przeoczy tego typu zagrania. Bezpieczniej będzie nie zwrócić id jeśli nie istnieje

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepsExecutor.cs:
```diff
@@ -7,19 +7,19 @@ namespace HotelReservation.ReservationSteps
 {
     public class StepsExecutor
```
To jest na obecną chwilę serce Twojego projektu, a nie posiada żadnego testu ;p Ta klasa wręcz musi być obtestowana. Jeśli z jakichś powodów jest ciężko coś przetestować to jest hint, że należy klasę lekko przeorganizować.

DanielDuchnicki replied:
Fixed - nie jestem pewien co do testowania wywołania stepu, który nie jest obsługiwany przez kod. Na tę chwilę dodałem TestStep do Enuma ReservationStepType.

michalborowski replied:
No i tutaj zaczyna się testowanie w izolacji.
ShouldNotThrowExceptionForValidReservationStepsList
ShouldThrowExceptionForNotValidReservationStepsList

To są testy, które dotyczą factory. Niech ona sobie rzuci wyjątek jak nie potrafi stworzyć stepu. Co zresztą robi, ale testu na to nie ma - dodaj proszę.
ReservationStepType.TestStep - coś takiego nie ma sensu w kodzie produkcyjnym, dlatego nie możemy dodać tego na potrzeby testów. Enum to jest pod spodem numerek. Można tego użyć do testowania przypadków, gdy jakaś wartość jest nieobsługiwana. Rzutowanie -1 na enum powinno wystaczyć:

_subject.CreateInstance((ReservationStepType)(-1)).Should().ThrowException()

Handlowanie wyjątków w StepExecutor nie jest jakoś super potrzebne wg mnie. Message w wyjątku jest bardziej dla programisty, niż dla usera. Tak naprawdę połykasz informacje przez co trudniej będzie się dowiedzieć gdzie wystąpił błąd.

Wracając do testów:
* Przypadki testowe mają odzwierciedlać odpowiedzialność klasy. Jakbym Cię zapytał co ta klasa robi, to powiesz, że rzuca wyjątek lub nie? Nie brakuje Ci tutaj nic? 
* Chcielibyśmy testować klasy w oderwaniu od innych klas. Obecnie testujesz tą klasę razem z kodem factory. Słowa kluczowe do poszukania: mock/stub

Jeśli będziesz miał problemy z przetestowaniem tej klasy w odizolowaniu od factory - zagadaj mnie to Ci trochę więcej podpowiem

DanielDuchnicki replied:
> No i tutaj zaczyna się testowanie w izolacji.
> ShouldNotThrowExceptionForValidReservationStepsList
> ShouldThrowExceptionForNotValidReservationStepsList
> 

> Wracając do testów:
> 
> * Przypadki testowe mają odzwierciedlać odpowiedzialność klasy. Jakbym Cię zapytał co ta klasa robi, to powiesz, że rzuca wyjątek lub nie? Nie brakuje Ci tutaj nic?
> * Chcielibyśmy testować klasy w oderwaniu od innych klas. Obecnie testujesz tą klasę razem z kodem factory. Słowa kluczowe do poszukania: mock/stub
> 
> Jeśli będziesz miał problemy z przetestowaniem tej klasy w odizolowaniu od factory - zagadaj mnie to Ci trochę więcej podpowiem

Reszta, poza wymienionym wyżej - fixed.

Zastanawiam się, czy użycie biblioteki "Moq" do stworzenia mocków/stubów jest najlepszym rozwiązaniem (dodałem ją już do projektu). Możemy chwilę o tym pogadać.

DanielDuchnicki committed:
- changing packages versions (FluentAssertions and NUnit)
- changing LastHotelId property
- changing GetHotelReservationSteps() method and user interface part that uses it
- adding StepFactory class
- changing existing tests and adding new ones

michalborowski commented:
> Mógłbyś w projekcie testowym zmienić wersje paczek na:
> 
> "FluentAssertions" version="5.3.0" targetFramework="net452" />
> "NUnit" version="3.11.0" targetFramework="net452" />
> 
> Pozwoli mi to łatwiej przeglądać w pracy ;D akturat nie mamy wersji, których użyłeś w internal repo

Jak zrobiłeś tę zmianę? Nugety mogą przy instalacji i usuwaniu wykonywać jakieś akcje. Masz w pliku csproj wpisy dotyczące innej wersji, więc coś poszło nie tak ;) Musisz zmienić wpisy (NUnit.3.12.0) na 3.11

DanielDuchnicki committed:
- changing LastHotelId property
- updates to StepFactory and StepsExecutor classes
- changing Nunit packages versions
- changes to tests

DanielDuchnicki commented:
> > Mógłbyś w projekcie testowym zmienić wersje paczek na:
> > "FluentAssertions" version="5.3.0" targetFramework="net452" />
> > "NUnit" version="3.11.0" targetFramework="net452" />
> > Pozwoli mi to łatwiej przeglądać w pracy ;D akturat nie mamy wersji, których użyłeś w internal repo
> 
> Jak zrobiłeś tę zmianę? Nugety mogą przy instalacji i usuwaniu wykonywać jakieś akcje. Masz w pliku csproj wpisy dotyczące innej wersji, więc coś poszło nie tak ;) Musisz zmienić wpisy (NUnit.3.12.0) na 3.11

Fixed (tak mi się wydaje)

DanielDuchnicki committed:
Changes to StepFactory and StepExecutor classes and StepExecutor class tests. Adding FakeItEasy package.
